ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java

```diff
@@ -867,8 +867,25 @@ public boolean delistResource(XAResource xaRes, int flags)
 					{
 						if (XAUtils.canOptimizeDelist(xaRes))
 						{
-							xaRes.end(info.xid(), XAResource.TMFAIL);
-							xaRes.rollback(info.xid());
+							boolean rollbackOnly = false;
+							try {
+								xaRes.end(info.xid(), XAResource.TMFAIL);
+							} catch (XAException e1) {
+								if ((e1.errorCode >= XAException.XA_RBBASE)
+										&& (e1.errorCode < XAException.XA_RBEND)) {
+									rollbackOnly = true;
+								}
+							}
+
+							try {
+								xaRes.rollback(info.xid());
+							} catch (XAException e1) {
+								if (rollbackOnly && e1.errorCode == XAException.XAER_NOTA) {
+									// The resource rolled back
+								} else {
+									throw e1;
+								}
```

Comment:
Maybe:
```suggestion
								if (!rollbackOnly && e1.errorCode != XAException.XAER_NOTA) {
									throw e1;
								}
```
WDYT?

Reply:
Or maybe this `if` has been thought like that because you wanted to specify `The resource rolled back`?

Reply:
I would like to point it out in a comment that when we get the XAER_NOTA when we had a valid XA_RB in end that we are expecting the resource has rolled back. I suppose I could reorganise it to have the comment anyway. I'll try that.